src/XamlX/IL/Emitters/PropertyAssignmentEmitter.cs

```diff
@@ -33,114 +32,279 @@ List<IXamlPropertySetter> ValidateAndGetSetters(XamlPropertyAssignmentNode an)
-                    if (!type.IsAssignableFrom(value.Type.GetClrType()))
-                    {
-                        codeGen
-                            .Dup()
-                            .Isinst(type)
-                            .Brfalse(Next());
-                        checkNext = true;
-                    }
+            // If the value is a value type, always use the first one
+            if (valueType.IsValueType)
+            {
+                setters.RemoveRange(1, setters.Count - 1);
+                return;
+            }
 
-                    if (checkNext)
-                        hadJumps = true;
-                    
-                    ILEmitHelpers.EmitConvert(context, codeGen, value, value.Type.GetClrType(), type);
-                    context.Emit(setter, codeGen);
-                    if (hadJumps)
-                    {
-                        codeGen.Br(exit);
-                    }
+            for (int index = 0; index < setters.Count;)
+            {
+                var setter = setters[index];
+                var type = setter.Parameters.Last();
 
-                    if(!checkNext)
-                        break;
+                // the value is directly assignable by downcast and the setter allows null: it will always match
+                if (type.IsAssignableFrom(valueType) && setter.BinderParameters.AllowRuntimeNull)
+                {
+                    setters.RemoveRange(index + 1, setters.Count - index - 1);
+                    return;
+                }
+
+                // the value type has already been handled by a previous setter
+                if (setters.Take(index).Any(previous => IsAssignableToWithNullability(setter, previous)))
+                {
+                    setters.RemoveAt(index);
+                    continue;
                 }
 
+                ++index;
+            }
+        }
+
+        private static bool IsAssignableToWithNullability(IXamlPropertySetter from, IXamlPropertySetter to)
+            => to.Parameters.Last().IsAssignableFrom(from.Parameters.Last())
+               && (to.BinderParameters.AllowRuntimeNull || !from.BinderParameters.AllowRuntimeNull);
+
+        private static IXamlMethod GetOrCreateDynamicSetterMethod(
+            IXamlType parentType,
+            IReadOnlyList<IXamlType> valueTypes,
+            IReadOnlyList<IXamlPropertySetter> setters,
+            XamlEmitContextWithLocals<IXamlILEmitter, XamlILNodeEmitResult> context)
+        {
+            if (!context.TryGetItem(out DynamicSettersCache cache))
+            {
+                var settersType = context.CreateSubType(
+                    "DynamicSetters_" + context.Configuration.IdentifierGenerator.GenerateIdentifierPart(),
+                    context.Configuration.WellKnownTypes.Object);
+                cache = new DynamicSettersCache(settersType);
+                context.SetItem(cache);
```

Comment:
Since `DynamicSettersCache` actually holds an emitted type that could and should be reused when compiling other XAML files, I think we should modify the context to allow for cross-context item sharing.

e. g. `SetItem` would become `public void SetItem<T>(T item, bool reusable = false)` and context would use a shared dictionary if it gets one in constructor.

That would only work for Cecil backend since with SRE type would be finalized by the next line, but that would still remove some code duplication where we care about it.

Reply:
I agree. Currently even in the same XAML file the cache will be different per sub type (happens on deferred content, `XamlClosure` type generation), which indeed isn't ideal. 

But even with the context change, the context currently only exposes a way to define a new sub type, not access the current type or the defining module which would be necessary to define a new top-level type containing all the setters. Otherwise there will be access problems since the nested type `XamlClosure_X+XamlClosure_Y` might try to call a method on `XamlClosure_A+DynamicSetters` generated earlier (nested closure types are private).

Reply:
Now that Avalonia has lazy resources, the setters are almost never reused :( Is it ok if I tackle this problem in another PR after this one? Are breaking changes to the emit context acceptable?

Reply:
Since XamlX in its current state is supposed to be shipped as a git submodule (we don't have nuget packages), there is room for breaking API changes, yes.